File: online-judge/4th-semester/hw4/610_StreetDirections.py
from sys import stdin
from sys import setrecursionlimit 
setrecursionlimit(10000)

processed = []
dEdges = set()
graph = []
visited = []
low = []
father = []
discTime = 0

def bridgesAux(node):
    global discTime, visited, father, low, dEdges
    discTime += 1
    visited[node] = discTime
    low[node] = discTime
    
    for i in range(len(graph[node])):
        adjNode = graph[node][i]
        pair = (min(node, adjNode), max(node, adjNode))

        if pair not in processed:
            processed.append(pair)

            if visited[adjNode] == -1:
                father[adjNode] = node
                bridgesAux(adjNode)
                low[node] = min(low[node], low[adjNode])

                #Verificar si es un puente
                if low[adjNode] > visited[node]:
                    #añado el contrario
                    dEdges.add((adjNode, node))

            #back-edge
            elif adjNode != father[node]:
                low[node] = min(low[node], visited[adjNode])
            
            #añado el normal
            dEdges.add((node, adjNode))
   
def bridges():
    for i in range(len(graph)):
        if visited[i] == -1:
            bridgesAux(i)


def main():
    global graph, visited, low, father, discTime, processed, dEdges
    count = 1
    n, m = list(map(int, stdin.readline().split()))
    while(n != 0):
        processed = []
        dEdges = set()
        graph = [[] for _ in range(n)]
        visited = [-1 for _ in range(n)]
        low = [-1 for _ in range(n)]
        father = [-1 for _ in range(n)]
        discTime = 0

        for i in range(m):
            a, b = list(map(int, stdin.readline().split()))
            graph[a - 1].append(b - 1)
            graph[b - 1].append(a - 1)

        bridges()

        print(count)
        print()
        for p in sorted(dEdges):
            print(p[0] + 1, p[1] + 1)
        print("#")
        n, m = list(map(int, stdin.readline().split()))

        count += 1
        

main()

File: online-judge/4th-semester/hw4/610_StreetDirections.cpp
#include <iostream>
#include <vector>
#include <set>

using namespace std;

vector<pair<int, int>> processed; 
set<pair<int, int>> dEdges;
vector<vector<int>> graph;
int visited[1002];
int low[1002];
int father[1002];
int n, discTime = 0;

bool search(pair<int, int> p)
{
    for(int i = 0; i < processed.size(); i++)
    {
        if(processed[i].first == p.first && processed[i].second == p.second)
        {
            return true;
        }
    }
    return false;
}

// void bridgesAux(int node)
// {
//     int adjNode;
//     pair<int, int> p;
//     discTime ++;
//     visited[node] = discTime;
//     low[node] = discTime;

//     for(int i = 0; i < graph[node].size(); i++)
//     {
//         adjNode = graph[node][i];
//         p = make_pair(min(node, adjNode), max(node, adjNode));

//         if(!search(p))
//         {
//             processed.push_back(p);
            
//             if (visited[adjNode] == -1)
//             {
//                 father[adjNode] = node;
//                 bridgesAux(adjNode);
//                 low[node] = min(low[node], low[adjNode]);

//                 // Verificar si es un puente
//                 // añado el opuesto
//                 if(low[adjNode] > visited[node])
//                 {
//                     dEdges.insert(make_pair(adjNode, node));
//                 }
//             }
//             // back-edge
//             else if(adjNode != father[node])
//             {
//                 low[node] = min(low[node], visited[adjNode]);
//             }
            
//             // añado el normal
//             dEdges.insert(make_pair(node, adjNode));
//         }
//     }
// }

void bridgesAux(int node)
{
    int adjNode, tmp;
    discTime++;
    visited[node] = discTime;
    low[node] = discTime;

    for(int i = 0; i < graph[node].size(); i++)
    {
        adjNode = graph[node][i];

        if (visited[adjNode] == -1)
        {
            father[adjNode] = node;
            bridgesAux(adjNode);
            low[node] = min(low[node], low[adjNode]);
            
            // Verificar si es un puente
            // añado el opuesto
            if(low[adjNode] > visited[node])
            {
                dEdges.insert(make_pair(adjNode, node));
            }
            dEdges.insert(make_pair(node, adjNode));
            
        }
        // back-edge
        else if(adjNode != father[node])
        {
            tmp = low[node];
            low[node] = min(low[node], visited[adjNode]);
            if(low[node] != tmp)
            {
                //cout << node + 1 << " " << adjNode + 1 << endl;
                dEdges.insert(make_pair(node, adjNode));
            }

        }
    }
}


void bridges()
{
    for(int i = 0; i < n; i++)
    {
        if(visited[i] == -1)
        {
            bridgesAux(i);
        }
    }  
}



int main()
{
    int m, a, b, count = 1;
    cin >> n >> m;
    while(n != 0)
    {
        processed.clear();
        dEdges.clear();
        graph.clear();
        discTime = 0;
        for(int i = 0; i < n; i++)
        {
            graph.push_back(vector<int>());
            visited[i] = -1;
            father[i] = -1;
            low[i] = -1;
        }

        for(int i = 0; i < m; i++)
        {
            cin >> a >> b;
            graph[a - 1].push_back(b - 1);
            graph[b - 1].push_back(a - 1);
        }

        bridges();

        cout << count << "\n" << endl;
        for(set<pair<int, int>>::iterator it = dEdges.begin(); it != dEdges.end(); it++)
        {
            printf("%d %d\n", it->first + 1, it->second + 1);
        }
        cout << "#" << endl;
        cin >> n >> m;

        count ++;
    }
}      
